k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,188 @@
+package io.buoyant.k8s
+
+import com.twitter.finagle.http.{Request, Response}
+import com.twitter.finagle.{Path, Service}
+import com.twitter.util.{Updatable, Activity, Var, Future}
+import io.buoyant.k8s
+
+/**
+ * Ingress resource names can be reused across namespaces.
+ * This tuple is to guarantee uniqueness
+ */
+case class NamespacedName(
+  namespace: Option[String],
+  name: Option[String]
+)
+
+case class IngressSpec(
+  name: Option[String],
+  fallbackBackend: Option[IngressPath] = None,
+  rules: Seq[IngressPath]
+)
+
+case class IngressPath(
+  host: Option[String] = None,
+  path: Option[Path] = None,
+  namespace: String,
+  svc: String,
+  port: String
+)
+
+object IngressCache {
+
+  type VarUp[T] = Var[T] with Updatable[T]
+
+  def mk(ns: Option[String], apiClient: Service[Request, Response]) =
+    new Ns[v1beta1.Ingress, v1beta1.IngressWatch, v1beta1.IngressList, IngressCache]() {
+      override protected def mkResource(name: Option[String]) = ns match {
+        case Some(ns) => v1beta1.Api(apiClient).withNamespace(ns).ingresses
+        case None => v1beta1.Api(apiClient).ingresses
+      }
+      override protected def mkCache(name: Option[String]) = new IngressCache(name)
+    }
+
+}
+
+/**
+ * IngressCache watches for ingress changes
+ * and checks incoming requests against cached ingress rules.
+ *
+ * @param namespace: The k8s namespace to filter on. If None, it watches all namespaces.
+ */
+
+class IngressCache(namespace: Option[String])
```

Comment:
I think this can be greatly simplified if we make namespace mandatory (or optional with a default value) instead of allowing the watching of all namespaces.  If we do this I think we no longer need to subclass `Ns` or keep a 2-level cache.  The result should be very very similar to K8sDtabStore.  

Rather than exposing an `Activity[Map[NamespacedName, VarUp[IngressSpec]]]` we could simply expose an `Activity[IngressSpec]`.